transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,419 @@
+                    .0
+                    .item;
+
+                trace!(from = "direct", "yield item");
+
+                #[cfg(all(feature = "prometheus", not(test)))]
+                METRIC_QUEUE_SIZE.decrement(1.0f64);
+
+                return Poll::Ready(Some(data));
+            }
+
+            if let Some(waker) = channel.waker.as_mut() {
+                waker.clone_from(cx.waker());
+            } else {
+                let waker = cx.waker().clone();
+                channel.waker = Some(waker);
+            }
+
+            if let Some(next) = channel.buffer.peek() {
+                let remaining = next.0.release_at.duration_since(now);
+
+                trace!("reseting the timer");
+                channel.timer.reset(remaining);
+
+                let timer = std::pin::pin!(&mut channel.timer);
+                let res = timer.poll(cx);
+
+                return match res {
+                    Poll::Ready(_) => {
+                        trace!(from = "timer", "yield item");
+
+                        #[cfg(all(feature = "prometheus", not(test)))]
+                        METRIC_QUEUE_SIZE.decrement(1.0f64);
+
+                        Poll::Ready(Some(
+                            channel
+                                .buffer
+                                .pop()
+                                .expect("The value should be present within the locked access")
+                                .0
+                                .item,
+                        ))
+                    }
+                    Poll::Pending => {
+                        trace!(from = "timer", "pending");
+                        Poll::Pending
+                    }
+                };
+            }
+
+            trace!(from = "direct", "pending");
+            Poll::Pending
+        } else {
+            channel.receiver_active = false;
+            Poll::Ready(None)
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Receive a single delayed mixed item.
+    pub async fn recv(&mut self) -> Option<T> {
+        poll_fn(|cx| self.poll_next_unpin(cx)).await
+    }
+}
+
+/// Instantiate a mixing channel and return the sender and receiver end of the channel.
+pub fn channel<T>(cfg: crate::config::MixerConfig) -> (Sender<T>, Receiver<T>) {
+    #[cfg(all(feature = "prometheus", not(test)))]
+    {
+        // Initialize the lazy statics here
+        lazy_static::initialize(&METRIC_QUEUE_SIZE);
+        lazy_static::initialize(&METRIC_MIXER_AVERAGE_DELAY);
+    }
+
+    let mut buffer = BinaryHeap::new();
+    buffer.reserve(cfg.capacity);
+
+    let channel = Arc::new(Mutex::new(Channel::<T> {
+        buffer,
+        timer: Delay::new(Duration::from_secs(0)),
+        waker: None,
+        sender_count: 1,
+        receiver_active: true,
+        cfg,
+    }));
+    (
+        Sender {
+            channel: channel.clone(),
+        },
+        Receiver { channel },
+    )
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Still missing the test with `SinkExt::send` and `SinkExt::feed/flush`. 

I know we might not be using it in this context, but it is a good test that the mixer's Sink interface is sane.

Reply:
I see it doing:
````
for i in input.iter() {
            tx.send(*i)?;
}
````
Which is the synchronous `send`, not the one from `SinkExt`, which needs to be awaited.

Or am I looking somewhere wrong?

Reply:
I will rewrite it as async, any other part of the `SinkExt` I should test with that?

Reply:
Just one test with `SinkExt::send` and maybe one test which feeds multiple elements via `SinkExt::feed` and then flushes them via `SinkExt::flush`.
Both are using slightly different strategies when dealing with your `Sink`, so it's better to have them tested both.